src/sagemaker/serve/utils/hardware_detector.py

```diff
@@ -0,0 +1,102 @@
+# Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License"). You
+# may not use this file except in compliance with the License. A copy of
+# the License is located at
+#
+#     http://aws.amazon.com/apache2.0/
+#
+# or in the "license" file accompanying this file. This file is
+# distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF
+# ANY KIND, either express or implied. See the License for the specific
+# language governing permissions and limitations under the License.
+"""Utilites for identifying and analyzing instance gpu hardware"""
+from __future__ import absolute_import
+
+import logging
+
+from sagemaker import Session
+
+logger = logging.getLogger(__name__)
+
+
+fallback_gpu_resource_mapping = {
+    "ml.p5.48xlarge": {"Count": 8, "TotalGpuMemoryInMiB": 655360},
+    "ml.p4d.24xlarge": {"Count": 8, "TotalGpuMemoryInMiB": 327680},
+    "ml.p4de.24xlarge": {"Count": 8, "TotalGpuMemoryInMiB": 610352},
+    "ml.p3.2xlarge": {"Count": 1, "TotalGpuMemoryInMiB": 16384},
+    "ml.p3.8xlarge": {"Count": 4, "TotalGpuMemoryInMiB": 65536},
+    "ml.p3.16xlarge": {"Count": 8, "TotalGpuMemoryInMiB": 131072},
+    "ml.p3dn.24xlarge": {"Count": 8, "TotalGpuMemoryInMiB": 262144},
+    "ml.p2.xlarge": {"Count": 1, "TotalGpuMemoryInMiB": 12288},
+    "ml.p2.8xlarge": {"Count": 8, "TotalGpuMemoryInMiB": 98304},
+    "ml.p2.16xlarge": {"Count": 16, "TotalGpuMemoryInMiB": 196608},
+    "ml.g4dn.xlarge": {"Count": 1, "TotalGpuMemoryInMiB": 16384},
+    "ml.g4dn.2xlarge": {"Count": 1, "TotalGpuMemoryInMiB": 16384},
+    "ml.g4dn.4xlarge": {"Count": 1, "TotalGpuMemoryInMiB": 16384},
+    "ml.g4dn.8xlarge": {"Count": 1, "TotalGpuMemoryInMiB": 16384},
+    "ml.g4dn.16xlarge": {"Count": 1, "TotalGpuMemoryInMiB": 16384},
+    "ml.g4dn.12xlarge": {"Count": 4, "TotalGpuMemoryInMiB": 65536},
+    "ml.g5n.xlarge": {"Count": 1, "TotalGpuMemoryInMiB": 22888},
+    "ml.g5.2xlarge": {"Count": 1, "TotalGpuMemoryInMiB": 24576},
+    "ml.g5.4xlarge": {"Count": 1, "TotalGpuMemoryInMiB": 24576},
+    "ml.g5.8xlarge": {"Count": 1, "TotalGpuMemoryInMiB": 24576},
+    "ml.g5.16xlarge": {"Count": 1, "TotalGpuMemoryInMiB": 24576},
+    "ml.g5.12xlarge": {"Count": 4, "TotalGpuMemoryInMiB": 98304},
+    "ml.g5.24xlarge": {"Count": 4, "TotalGpuMemoryInMiB": 98304},
+    "ml.g5.48xlarge": {"Count": 8, "TotalGpuMemoryInMiB": 196608},
+}
+
+
+def _get_gpu_info(instance_type: str, session: Session) -> int:
```

Comment:
lets collapse these into one method. We can get just get the gpu count and memory from the dictionary at the place of calling

same comment for the fall back methods

ie:
```
_get_gpu_info(...) -> dict
_get_gpu_info_fallback(...) -> dict
```

Reply:
or filter down the dictionary to what you want and return that instead. maybe as a tuple or smth

Reply:
Agree return a tuple here of {gpu_count, total_gpu_memory_size_MB}